protocol/daemons/server/types/health_monitor.go

```diff
@@ -0,0 +1,277 @@
+package types
+
+import (
+	cosmoslog "cosmossdk.io/log"
+	"fmt"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/types"
+	libtime "github.com/dydxprotocol/v4-chain/protocol/lib/time"
+	"sync"
+	"time"
+)
+
+const (
+	HealthCheckPollFrequency = 5 * time.Second
+)
+
+// timestampWithError couples a timestamp and error to make it easier to update them in tandem.
+type timestampWithError struct {
+	timestamp time.Time
+	err       error
+}
+
+func (u *timestampWithError) Update(timestamp time.Time, err error) {
+	u.timestamp = timestamp
+	u.err = err
+}
+
+func (u *timestampWithError) Reset() {
+	u.Update(time.Time{}, nil)
+}
+
+func (u *timestampWithError) IsZero() bool {
+	return u.timestamp.IsZero()
+}
+
+func (u *timestampWithError) Timestamp() time.Time {
+	return u.timestamp
+}
+
+func (u *timestampWithError) Error() error {
+	return u.err
+}
+
+// healthChecker encapsulates the logic for monitoring the health of a health checkable service.
+type healthChecker struct {
+	// healthCheckable is the health checkable service to be monitored.
+	healthCheckable types.HealthCheckable
+
+	// timer triggers a health check poll for a health checkable service.
+	timer *time.Timer
+
+	// pollFrequency is the frequency at which the health checkable service is polled.
+	pollFrequency time.Duration
+
+	// mostRecentSuccess is the timestamp of the most recent successful health check.
+	mostRecentSuccess time.Time
+
+	// firstFailureInStreak is the timestamp of the first error in the most recent streak of errors. It is set
+	// whenever the service toggles from healthy to an unhealthy state, and used to determine how long the daemon has
+	// been unhealthy. If this timestamp is nil, then the error streak ended before it could trigger a callback.
+	firstFailureInStreak timestampWithError
```

Comment:
I think that I will leave it in order to avoid the potential issue of memory leaks for cases when someone disables the monitor by configuring it with a >> maximum unhealthy duration.

Reply:
We want to to keep track of the latest failure since we want to know if something has changed and the time at which the first failure occurred.

Storing only the first failure error will not allow us to see that the error condition has changed over time.